python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -27,6 +28,54 @@
 MaybeRefBundle = Union[RefBundle, Exception, None]
 
 
+@dataclass
+class TopologyResourceUsage:
+    """Snapshot of resource usage in a `Topology` object.
+
+    The stats here can be computed on the fly from any `Topology`; this class
+    serves only a convenience wrapper to access the current usage snapshot.
+    """
+
+    # The current usage of the topology (summed across all operators and queues).
+    overall: ExecutionResources
+
+    # The downstream resource usage by operator.
+    downstream_memory_usage: Dict[PhysicalOperator, "DownstreamMemoryInfo"]
+
+    @staticmethod
+    def of(topology: Topology) -> "TopologyResourceUsage":
+        """Calculate the resource usage of the given topology."""
+        downstream_usage = {}
+        cur_usage = ExecutionResources(0, 0, 0)
+        # Iterate from last to first operator.
+        for op, state in list(topology.items())[::-1]:
+            cur_usage = cur_usage.add(op.current_resource_usage())
+            # Don't count input refs towards dynamic memory usage, as they have been
+            # pre-created already outside this execution.
+            if not isinstance(op, InputDataBuffer):
+                for bundle in state.outqueue:
+                    cur_usage.object_store_memory += bundle.size_bytes()
+            # Subtract one from denom to account for input buffer.
+            f = (1.0 + len(downstream_usage)) / max(1.0, len(topology) - 1.0)
```

Comment:
Shall we add a scaling factor (5x?) for all-to-all operators for safety?

Reply:
The memory usage for AllToAll operators is not managed, so it's not relevant here.

Reply:
Oh, you mean the intermediate blocks in an AllToAll op is unaccounted, only its inputs and outputs?

I still have a question about memory safety: IIUC, in the worst case, if there are N operators, it can use up to (N+1)/2 times of the object store memory limit (from the last to the first op 1/N + 2/N + ... + N/N). Should we have a way to cap this risk?

Reply:
Yeah, we don't account for that at all (and it's not really meaningful to limit anyways, since it will always use the same amount of memory).

> I still have a question about memory safety: IIUC, in the worst case, if there are N operators, it can use up to (N+1)/2 times of the object store memory limit (from the last to the first op 1/N + 2/N + ... + N/N). Should we have a way to cap this risk?

The backpressure algorithm will already seek to schedule operators in a way that avoids this scenario. The only case when it can happen is there is a big spike in memory at the start. In that case, there's no better alternative that to temporarily allow extra memory usage to clear the backlog--- we can explore avoiding memory spikes in the future with slow start algorithms, better estimation, etc.